### PR TITLE
Terraform Restore Improvements for GKEBackup

### PR DESCRIPTION
**Pull request by avijitagarwal195**
Adding changes to support restore improvements to Backup For GKE. 
Fixes b/327225667

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
gkebackup: added `backup_schedule.backup_config.permissive_mode` field to `google_gke_backup_backup_plan` resource
```
```release-note:enhancement
gkebackup: added `restore_config.restore_order` field to `google_gke_backup_restore_plan` resource
```
```release-note:enhancement
gkebackup: added `restore_config.volume_data_restore_policy_bindings` field to `google_gke_backup_restore_plan` resource
```
```release-note:enhancement
gkebackup: added new enum values `MERGE_SKIP_ON_CONFLICT`, `MERGE_REPLACE_VOLUME_ON_CONFLICT` and `MERGE_REPLACE_ON_CONFLICT` to field `restore_config.namespaced_resource_restore_mode` in `google_gke_backup_restore_plan` resource
```

### CODE REVIEW

**Comment by avijitagarwal195** [comment]
The newly added tests are failing in this PR  because the new fields are hidden behind API visibility label. I will wait till the visibility labels are removed and pushed to production and then re-trigger the tests in CI. The ETA for visibility labels to be removed in mid of May.

**Comment by zli82016** [comment]
Feel free to let me know when the PR is ready for review.

**Comment by avijitagarwal195** [comment]
@Zhenhua Li This PR is ready for review now. The visibility labels have been removed and the added acceptance tests are running successfully now. Thank you.

File: mmv1/products/gkebackup/BackupPlan.yaml [hunk 1]
```diff
@@ -441,6 +458,13 @@ properties:
                   required: true
                   description: |
                     The name of a Kubernetes Resource.
+      - !ruby/object:Api::Type::Boolean
+        name: permissiveMode
+        description: |
+          This flag specifies whether Backups will not fail when
+          Backup for GKE detects Kubernetes configuration that is
+          non-standard or requires additional setup to restore.
+        default_from_api: false
```
Comment (zli82016) [inline mmv1/products/gkebackup/BackupPlan.yaml:467]: `default_from_api: false` is not needed here, as it the default value is false.

File: mmv1/products/gkebackup/BackupPlan.yaml [hunk 2]
```diff
@@ -441,6 +458,13 @@ properties:
                   required: true
                   description: |
                     The name of a Kubernetes Resource.
+      - !ruby/object:Api::Type::Boolean
```
Comment (zli82016) [inline mmv1/products/gkebackup/BackupPlan.yaml:461]: Can you please add the update test for this field?
I recommend first setting this field to true, and then removing this field in the update test steps.

 The doc for the update test is [here](https://googlecloudplatform.github.io/magic-modules/develop/test/test/#add-an-update-test)

File: mmv1/products/gkebackup/RestorePlan.yaml [hunk 3]
```diff
@@ -496,6 +547,87 @@ properties:
...
+                - :REUSE_VOLUME_HANDLE_FROM_BACKUP
+                - :NO_VOLUME_DATA_RESTORATION
+            - !ruby/object:Api::Type::Enum
+              name: volumeType
+              required: true
+              description: |
+                The volume type, as determined by the PVC's
+                bound PV, to apply the policy to.
+              exactly_one_of:
+                - restoreConfig.0.volumeType
```
Comment (zli82016) [inline mmv1/products/gkebackup/RestorePlan.yaml:577]: Is `exactly_one_of` needed here?

File: mmv1/products/gkebackup/RestorePlan.yaml [hunk 4]
```diff
@@ -496,6 +547,87 @@ properties:
...
+              required: true
+              properties:
+                - !ruby/object:Api::Type::NestedObject
+                  name: satisfying
+                  required: true
+                  description: |
+                    The satisfying group kind must be restored first
+                    in order to satisfy the dependency.
+                  properties:
+                    - !ruby/object:Api::Type::String
```
Comment (zli82016) [inline mmv1/products/gkebackup/RestorePlan.yaml:602]: If `resourceGroup` is not specified in the configuration and not included in the request to API, will it default to core group in the API side in the response?

File: mmv1/products/gkebackup/RestorePlan.yaml [hunk 5]
```diff
@@ -496,6 +547,87 @@ properties:
                     description: |
                       A string that specifies the desired value in string format
                       to use for transformation.
+      - !ruby/object:Api::Type::Array
```
Comment (zli82016) [inline mmv1/products/gkebackup/RestorePlan.yaml:550]: Can you also add update test for these new fields? Thanks.

**Review by zli82016 — CHANGES_REQUESTED** [review]

File: mmv1/products/gkebackup/RestorePlan.yaml [hunk 6]
```diff
@@ -496,6 +547,87 @@ properties:
...
+                  description: |
+                    The satisfying group kind must be restored first
+                    in order to satisfy the dependency.
+                  properties:
+                    - !ruby/object:Api::Type::String
+                      name: resourceGroup
+                      description: |
+                        API Group of a Kubernetes resource, e.g.
+                        "apiextensions.k8s.io", "storage.k8s.io", etc.
+                        Use empty string for core group.
```
Comment (zli82016) [inline mmv1/products/gkebackup/RestorePlan.yaml:607]: If `resourceGroup` is set to empty string "" in the configuration and included in the request to API, what value will be in the API response?

File: mmv1/products/gkebackup/BackupPlan.yaml (hunk 1 above)
Comment (avijitagarwal195) [reply mmv1/products/gkebackup/BackupPlan.yaml:467]: Done

File: mmv1/products/gkebackup/BackupPlan.yaml (hunk 2 above)
Comment (avijitagarwal195) [reply mmv1/products/gkebackup/BackupPlan.yaml:461]: Thanks Zhenhua! Added. PTAL.

File: mmv1/products/gkebackup/RestorePlan.yaml (hunk 5 above)
Comment (avijitagarwal195) [reply mmv1/products/gkebackup/RestorePlan.yaml:550]: Added. PTAL.

File: mmv1/products/gkebackup/RestorePlan.yaml (hunk 3 above)
Comment (avijitagarwal195) [reply mmv1/products/gkebackup/RestorePlan.yaml:577]: Ahh yes this is no needed. Removed. Thank you!

File: mmv1/products/gkebackup/RestorePlan.yaml (hunk 4 above)
Comment (avijitagarwal195) [reply mmv1/products/gkebackup/RestorePlan.yaml:602]: We can leave this empty. Internally in the backend we would infer empty as a core group so it does not need to have any resourceGroup value for it.

File: mmv1/products/gkebackup/RestorePlan.yaml (hunk 6 above)
Comment (avijitagarwal195) [reply mmv1/products/gkebackup/RestorePlan.yaml:607]: Leaving this as empty string will also make this behave as a core group as empty string is the default value of string type (go/proto3#default)

**Review by zli82016 — CHANGES_REQUESTED** [review]
LGTM. Thanks.

**Review by zli82016 — APPROVED** [review]
LGTM.

**Pull request merged**